{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/powerapps/share/python-anaconda-2019.10/lib/python3.7/site-packages/requests/__init__.py:91: RequestsDependencyWarning: urllib3 (1.26.13) or chardet (3.0.4) doesn't match a supported version!\n",
      "  RequestsDependencyWarning)\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import io\n",
    "import sys\n",
    "import toytree\n",
    "import toyplot.svg\n",
    "import pandas as pd \n",
    "from Bio import SeqIO, AlignIO\n",
    "from Bio.Align.Applications import MafftCommandline\n",
    "sys.path.append(r'/davidb/yatirsolan/scripts/python/bio_utilities')\n",
    "sys.path.append(r'/davidb/yatirsolan/thesis_work/figures/general')\n",
    "import computational_tools\n",
    "import phylogenetics\n",
    "import databases\n",
    "import HMM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "working_directory = r'/davidb/yatirsolan/thesis_work/figures/novel_insights/T4SSB_figure/'\n",
    "full_genomes_file = 'species_full_genomes.faa'\n",
    "full_genomes_dict = SeqIO.to_dict(SeqIO.parse(os.path.join(working_directory, full_genomes_file), 'fasta'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "! /davidb/local/bin/hmmsearch -o /dev/null --tblout T4SSB_HMMs_vs_species_full_genomes.hmmsearch --notextw -E 1e-4 --cpu 4 /davidb/yatirsolan/secretion_systems/T4SS/hmm_repository/T4SSB.Hmm /davidb/yatirsolan/thesis_work/figures/novel_insights/T4SSB_figure/species_full_genomes.faa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/davidb/yatirsolan/thesis_work/figures/novel_insights/T4SSB_figure/T4SSB_HMMs_vs_species_full_genomes.hmmsearch.tsv'"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "HMM.hmmsearch_to_tsv(r'/davidb/yatirsolan/thesis_work/figures/novel_insights/T4SSB_figure/T4SSB_HMMs_vs_species_full_genomes.hmmsearch')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "GCA_to_taxid_dict = dict(pd.read_table(r'/davidb/bio_db/NCBI/WGS/latest.assembly_summary_genbank.wLineage.no_fungi_metazoa_viridiplantae.tab', usecols=['# assembly_accession', 'taxid']).values)\n",
    "GCA_to_taxid_dict = {k:str(v) for k, v in GCA_to_taxid_dict.items()}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_to_include = ['Legionella oakridgensis',\n",
    "                      'Legionella erythra',\n",
    "                      'Legionella hackeliae',\n",
    "                      'Legionella pneumophila subsp. pneumophila str. Philadelphia 1',\n",
    "                      'Legionella moravica',\n",
    "                      'Legionella anisa',\n",
    "                      'Tatlockia micdadei',\n",
    "                      'Fluoribacter dumoffii NY 23',\n",
    "                      'Legionella longbeachae',\n",
    "                      'Piscirickettsia salmonis',\n",
    "                      'Piscirickettsia litoralis',\n",
    "                      'Pseudomonas amygdali pv. sesami',\n",
    "                      'Pseudomonas monteilii',\n",
    "                      'Pseudomonas parafulva NBRC 16636 = DSM 17004',\n",
    "                      'Pseudomonas putida B001',\n",
    "                      'Coxiella burnetii RSA 493',\n",
    "                      'Rickettsiella grylli',\n",
    "                      'Legionella geestiana']\n",
    "\n",
    "species_to_exclude = []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>targetName</th>\n",
       "      <th>queryName</th>\n",
       "      <th>queryAccession</th>\n",
       "      <th>E-value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Legionella_shakespearei_DSM_23087|SAMN04274794...</td>\n",
       "      <td>DotA</td>\n",
       "      <td>Leg_DotA</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Legionella_moravica|SAMN04274785|GCA_001467865...</td>\n",
       "      <td>DotA</td>\n",
       "      <td>Leg_DotA</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Fluoribacter_bozemanae|SAMN04274766|GCA_001467...</td>\n",
       "      <td>DotA</td>\n",
       "      <td>Leg_DotA</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Legionella_parisiensis|SAMN04274788|GCA_001467...</td>\n",
       "      <td>DotA</td>\n",
       "      <td>Leg_DotA</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Fluoribacter_dumoffii_NY_23|SAMN04274771|GCA_0...</td>\n",
       "      <td>DotA</td>\n",
       "      <td>Leg_DotA</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7217</th>\n",
       "      <td>Pseudomonas_amygdali_pv._sesami|SAMN03976303|G...</td>\n",
       "      <td>DotC</td>\n",
       "      <td>Pisc_DotC</td>\n",
       "      <td>1.500000e-30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7218</th>\n",
       "      <td>Pseudomonas_monteilii|SAMN11579699|GCA_0098313...</td>\n",
       "      <td>DotC</td>\n",
       "      <td>Pisc_DotC</td>\n",
       "      <td>8.000000e-30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7219</th>\n",
       "      <td>Pseudomonas_putida_B001|SAMEA2272788|GCA_00028...</td>\n",
       "      <td>DotC</td>\n",
       "      <td>Pisc_DotC</td>\n",
       "      <td>8.000000e-30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7220</th>\n",
       "      <td>Piscirickettsia_litoralis|SAMN05601431|GCA_001...</td>\n",
       "      <td>DotC</td>\n",
       "      <td>Pisc_DotC</td>\n",
       "      <td>2.200000e-27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7221</th>\n",
       "      <td>Piscirickettsia_litoralis|SAMN05601431|GCA_001...</td>\n",
       "      <td>DotC</td>\n",
       "      <td>Pisc_DotC</td>\n",
       "      <td>2.400000e-17</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7222 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             targetName queryName  \\\n",
       "0     Legionella_shakespearei_DSM_23087|SAMN04274794...      DotA   \n",
       "1     Legionella_moravica|SAMN04274785|GCA_001467865...      DotA   \n",
       "2     Fluoribacter_bozemanae|SAMN04274766|GCA_001467...      DotA   \n",
       "3     Legionella_parisiensis|SAMN04274788|GCA_001467...      DotA   \n",
       "4     Fluoribacter_dumoffii_NY_23|SAMN04274771|GCA_0...      DotA   \n",
       "...                                                 ...       ...   \n",
       "7217  Pseudomonas_amygdali_pv._sesami|SAMN03976303|G...      DotC   \n",
       "7218  Pseudomonas_monteilii|SAMN11579699|GCA_0098313...      DotC   \n",
       "7219  Pseudomonas_putida_B001|SAMEA2272788|GCA_00028...      DotC   \n",
       "7220  Piscirickettsia_litoralis|SAMN05601431|GCA_001...      DotC   \n",
       "7221  Piscirickettsia_litoralis|SAMN05601431|GCA_001...      DotC   \n",
       "\n",
       "     queryAccession       E-value  \n",
       "0          Leg_DotA  0.000000e+00  \n",
       "1          Leg_DotA  0.000000e+00  \n",
       "2          Leg_DotA  0.000000e+00  \n",
       "3          Leg_DotA  0.000000e+00  \n",
       "4          Leg_DotA  0.000000e+00  \n",
       "...             ...           ...  \n",
       "7217      Pisc_DotC  1.500000e-30  \n",
       "7218      Pisc_DotC  8.000000e-30  \n",
       "7219      Pisc_DotC  8.000000e-30  \n",
       "7220      Pisc_DotC  2.200000e-27  \n",
       "7221      Pisc_DotC  2.400000e-17  \n",
       "\n",
       "[7222 rows x 4 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hmmsrch_df = pd.read_table(os.path.join(working_directory, 'T4SSB_HMMs_vs_species_full_genomes.hmmsearch.tsv'), \n",
    "                           usecols=['targetName', 'queryName', 'queryAccession', 'E-value'])\n",
    "hmmsrch_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>targetName</th>\n",
       "      <th>queryName</th>\n",
       "      <th>queryAccession</th>\n",
       "      <th>E-value</th>\n",
       "      <th>GCA</th>\n",
       "      <th>tax</th>\n",
       "      <th>seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2829</th>\n",
       "      <td>Coxiella_burnetii_RSA_493|SAMN02603970|GCA_000...</td>\n",
       "      <td>DotA</td>\n",
       "      <td>Cox_DotA</td>\n",
       "      <td>1.100000e-280</td>\n",
       "      <td>GCA_000007765.2</td>\n",
       "      <td>Coxiella burnetii RSA 493</td>\n",
       "      <td>(M, K, K, L, V, S, S, L, L, A, S, I, S, L, F, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Fluoribacter_dumoffii_NY_23|SAMN04274771|GCA_0...</td>\n",
       "      <td>DotA</td>\n",
       "      <td>Leg_DotA</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>GCA_001467605.1</td>\n",
       "      <td>Fluoribacter dumoffii NY 23</td>\n",
       "      <td>(M, K, R, L, L, G, T, L, L, L, L, L, F, P, G, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Legionella_anisa|SAMN04274764|GCA_001467525.1|...</td>\n",
       "      <td>DotA</td>\n",
       "      <td>Leg_DotA</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>GCA_001467525.1</td>\n",
       "      <td>Legionella anisa</td>\n",
       "      <td>(M, N, K, L, L, V, T, L, L, L, L, V, C, P, G, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Legionella_erythra|SAMN04274772|GCA_001467615....</td>\n",
       "      <td>DotA</td>\n",
       "      <td>Leg_DotA</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>GCA_001467615.1</td>\n",
       "      <td>Legionella erythra</td>\n",
       "      <td>(M, N, K, I, V, S, T, L, L, M, T, L, F, P, A, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>Legionella_geestiana|SAMEA103899863|GCA_900452...</td>\n",
       "      <td>DotA</td>\n",
       "      <td>Leg_DotA</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>GCA_900452375.1</td>\n",
       "      <td>Legionella geestiana</td>\n",
       "      <td>(M, K, K, Y, I, L, G, L, L, S, A, L, M, P, V, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2039</th>\n",
       "      <td>Legionella_moravica|SAMN04274785|GCA_001467865...</td>\n",
       "      <td>IcmX</td>\n",
       "      <td>Leg_IcmX</td>\n",
       "      <td>5.400000e-184</td>\n",
       "      <td>GCA_001467865.1</td>\n",
       "      <td>Legionella moravica</td>\n",
       "      <td>(M, K, L, L, S, K, C, S, L, M, T, L, L, C, L, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2068</th>\n",
       "      <td>Legionella_oakridgensis|SAMN04274787|GCA_00146...</td>\n",
       "      <td>IcmX</td>\n",
       "      <td>Leg_IcmX</td>\n",
       "      <td>1.900000e-133</td>\n",
       "      <td>GCA_001467925.1</td>\n",
       "      <td>Legionella oakridgensis</td>\n",
       "      <td>(M, K, I, I, G, R, S, L, F, A, C, L, L, S, M, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2058</th>\n",
       "      <td>Legionella_pneumophila_subsp._pneumophila_str....</td>\n",
       "      <td>IcmX</td>\n",
       "      <td>Leg_IcmX</td>\n",
       "      <td>7.900000e-167</td>\n",
       "      <td>GCA_000008485.1</td>\n",
       "      <td>Legionella pneumophila subsp. pneumophila str....</td>\n",
       "      <td>(M, K, V, L, P, K, L, A, L, A, N, L, V, C, F, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4735</th>\n",
       "      <td>Rickettsiella_grylli|SAMN02435840|GCA_00016829...</td>\n",
       "      <td>IcmX</td>\n",
       "      <td>Cox_IcmX</td>\n",
       "      <td>8.400000e-153</td>\n",
       "      <td>GCA_000168295.1</td>\n",
       "      <td>Rickettsiella grylli</td>\n",
       "      <td>(M, M, Y, P, S, K, I, K, K, I, K, I, H, F, I, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2051</th>\n",
       "      <td>Tatlockia_micdadei|SAMN04274784|GCA_001467875....</td>\n",
       "      <td>IcmX</td>\n",
       "      <td>Leg_IcmX</td>\n",
       "      <td>1.600000e-171</td>\n",
       "      <td>GCA_001467875.1</td>\n",
       "      <td>Tatlockia micdadei</td>\n",
       "      <td>(M, K, S, L, S, R, I, V, L, P, G, V, L, F, F, ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>404 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             targetName queryName  \\\n",
       "2829  Coxiella_burnetii_RSA_493|SAMN02603970|GCA_000...      DotA   \n",
       "4     Fluoribacter_dumoffii_NY_23|SAMN04274771|GCA_0...      DotA   \n",
       "10    Legionella_anisa|SAMN04274764|GCA_001467525.1|...      DotA   \n",
       "8     Legionella_erythra|SAMN04274772|GCA_001467615....      DotA   \n",
       "37    Legionella_geestiana|SAMEA103899863|GCA_900452...      DotA   \n",
       "...                                                 ...       ...   \n",
       "2039  Legionella_moravica|SAMN04274785|GCA_001467865...      IcmX   \n",
       "2068  Legionella_oakridgensis|SAMN04274787|GCA_00146...      IcmX   \n",
       "2058  Legionella_pneumophila_subsp._pneumophila_str....      IcmX   \n",
       "4735  Rickettsiella_grylli|SAMN02435840|GCA_00016829...      IcmX   \n",
       "2051  Tatlockia_micdadei|SAMN04274784|GCA_001467875....      IcmX   \n",
       "\n",
       "     queryAccession        E-value              GCA  \\\n",
       "2829       Cox_DotA  1.100000e-280  GCA_000007765.2   \n",
       "4          Leg_DotA   0.000000e+00  GCA_001467605.1   \n",
       "10         Leg_DotA   0.000000e+00  GCA_001467525.1   \n",
       "8          Leg_DotA   0.000000e+00  GCA_001467615.1   \n",
       "37         Leg_DotA   0.000000e+00  GCA_900452375.1   \n",
       "...             ...            ...              ...   \n",
       "2039       Leg_IcmX  5.400000e-184  GCA_001467865.1   \n",
       "2068       Leg_IcmX  1.900000e-133  GCA_001467925.1   \n",
       "2058       Leg_IcmX  7.900000e-167  GCA_000008485.1   \n",
       "4735       Cox_IcmX  8.400000e-153  GCA_000168295.1   \n",
       "2051       Leg_IcmX  1.600000e-171  GCA_001467875.1   \n",
       "\n",
       "                                                    tax  \\\n",
       "2829                          Coxiella burnetii RSA 493   \n",
       "4                           Fluoribacter dumoffii NY 23   \n",
       "10                                     Legionella anisa   \n",
       "8                                    Legionella erythra   \n",
       "37                                 Legionella geestiana   \n",
       "...                                                 ...   \n",
       "2039                                Legionella moravica   \n",
       "2068                            Legionella oakridgensis   \n",
       "2058  Legionella pneumophila subsp. pneumophila str....   \n",
       "4735                               Rickettsiella grylli   \n",
       "2051                                 Tatlockia micdadei   \n",
       "\n",
       "                                                    seq  \n",
       "2829  (M, K, K, L, V, S, S, L, L, A, S, I, S, L, F, ...  \n",
       "4     (M, K, R, L, L, G, T, L, L, L, L, L, F, P, G, ...  \n",
       "10    (M, N, K, L, L, V, T, L, L, L, L, V, C, P, G, ...  \n",
       "8     (M, N, K, I, V, S, T, L, L, M, T, L, F, P, A, ...  \n",
       "37    (M, K, K, Y, I, L, G, L, L, S, A, L, M, P, V, ...  \n",
       "...                                                 ...  \n",
       "2039  (M, K, L, L, S, K, C, S, L, M, T, L, L, C, L, ...  \n",
       "2068  (M, K, I, I, G, R, S, L, F, A, C, L, L, S, M, ...  \n",
       "2058  (M, K, V, L, P, K, L, A, L, A, N, L, V, C, F, ...  \n",
       "4735  (M, M, Y, P, S, K, I, K, K, I, K, I, H, F, I, ...  \n",
       "2051  (M, K, S, L, S, R, I, V, L, P, G, V, L, F, F, ...  \n",
       "\n",
       "[404 rows x 7 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def edit_seq(seq):\n",
    "    seq.name = seq.name.split('|')[0]\n",
    "    seq.name = '_'.join(seq.name.split('_')[:2:])\n",
    "    seq.id = seq.id.split('|')[0]\n",
    "    seq.id = '_'.join(seq.id.split('_')[:2:])\n",
    "    seq.description = str()\n",
    "    return seq\n",
    "\n",
    "hmmsrch_df['GCA'] = hmmsrch_df.targetName.map(databases.gca_from_accession)\n",
    "hmmsrch_df['tax'] = hmmsrch_df.GCA.map(GCA_to_taxid_dict)\n",
    "hmmsrch_df['tax'] = hmmsrch_df.tax.map(phylogenetics.taxid_to_name)\n",
    "hmmsrch_df = hmmsrch_df[hmmsrch_df.tax.isin(species_to_include)]\n",
    "\n",
    "hmmsrch_df = hmmsrch_df.sort_values('E-value')\n",
    "hmmsrch_df = hmmsrch_df.drop_duplicates('targetName', keep='first')\n",
    "hmmsrch_df = hmmsrch_df.sort_values(['queryName', 'tax', 'E-value'])\n",
    "hmmsrch_df = hmmsrch_df.drop_duplicates(subset=['queryName', 'tax'], keep='first')\n",
    "hmmsrch_df = hmmsrch_df.sort_values(['queryName', 'tax'])\n",
    "\n",
    "hmmsrch_df['seq'] = hmmsrch_df.targetName.map(full_genomes_dict)\n",
    "hmmsrch_df['seq'] = hmmsrch_df.seq.map(edit_seq)\n",
    "\n",
    "hmmsrch_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>targetName</th>\n",
       "      <th>queryName</th>\n",
       "      <th>queryAccession</th>\n",
       "      <th>E-value</th>\n",
       "      <th>GCA</th>\n",
       "      <th>tax</th>\n",
       "      <th>seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2829</th>\n",
       "      <td>Coxiella_burnetii_RSA_493|SAMN02603970|GCA_000...</td>\n",
       "      <td>DotA</td>\n",
       "      <td>Cox_DotA</td>\n",
       "      <td>1.100000e-280</td>\n",
       "      <td>GCA_000007765.2</td>\n",
       "      <td>Coxiella burnetii RSA 493</td>\n",
       "      <td>(M, K, K, L, V, S, S, L, L, A, S, I, S, L, F, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Fluoribacter_dumoffii_NY_23|SAMN04274771|GCA_0...</td>\n",
       "      <td>DotA</td>\n",
       "      <td>Leg_DotA</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>GCA_001467605.1</td>\n",
       "      <td>Fluoribacter dumoffii NY 23</td>\n",
       "      <td>(M, K, R, L, L, G, T, L, L, L, L, L, F, P, G, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Legionella_anisa|SAMN04274764|GCA_001467525.1|...</td>\n",
       "      <td>DotA</td>\n",
       "      <td>Leg_DotA</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>GCA_001467525.1</td>\n",
       "      <td>Legionella anisa</td>\n",
       "      <td>(M, N, K, L, L, V, T, L, L, L, L, V, C, P, G, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Legionella_erythra|SAMN04274772|GCA_001467615....</td>\n",
       "      <td>DotA</td>\n",
       "      <td>Leg_DotA</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>GCA_001467615.1</td>\n",
       "      <td>Legionella erythra</td>\n",
       "      <td>(M, N, K, I, V, S, T, L, L, M, T, L, F, P, A, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>Legionella_geestiana|SAMEA103899863|GCA_900452...</td>\n",
       "      <td>DotA</td>\n",
       "      <td>Leg_DotA</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>GCA_900452375.1</td>\n",
       "      <td>Legionella geestiana</td>\n",
       "      <td>(M, K, K, Y, I, L, G, L, L, S, A, L, M, P, V, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4634</th>\n",
       "      <td>Pseudomonas_monteilii|SAMN11579699|GCA_0098313...</td>\n",
       "      <td>IcmT</td>\n",
       "      <td>Cox_IcmT</td>\n",
       "      <td>6.500000e-06</td>\n",
       "      <td>GCA_009831395.1</td>\n",
       "      <td>Pseudomonas monteilii</td>\n",
       "      <td>(M, A, T, G, R, Y, H, W, S, R, C, A, P, P, A, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4632</th>\n",
       "      <td>Pseudomonas_parafulva_NBRC_16636___DSM_17004|S...</td>\n",
       "      <td>IcmT</td>\n",
       "      <td>Cox_IcmT</td>\n",
       "      <td>1.100000e-06</td>\n",
       "      <td>GCA_000730645.1</td>\n",
       "      <td>Pseudomonas parafulva NBRC 16636 = DSM 17004</td>\n",
       "      <td>(M, A, T, G, R, Y, H, W, S, R, S, A, P, P, A, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4635</th>\n",
       "      <td>Pseudomonas_putida_B001|SAMEA2272788|GCA_00028...</td>\n",
       "      <td>IcmT</td>\n",
       "      <td>Cox_IcmT</td>\n",
       "      <td>6.500000e-06</td>\n",
       "      <td>GCA_000285395.1</td>\n",
       "      <td>Pseudomonas putida B001</td>\n",
       "      <td>(M, A, T, G, R, Y, H, W, S, R, C, A, P, P, A, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4589</th>\n",
       "      <td>Rickettsiella_grylli|SAMN02435840|GCA_00016829...</td>\n",
       "      <td>IcmT</td>\n",
       "      <td>Cox_IcmT</td>\n",
       "      <td>1.100000e-43</td>\n",
       "      <td>GCA_000168295.1</td>\n",
       "      <td>Rickettsiella grylli</td>\n",
       "      <td>(M, S, R, V, P, T, A, A, H, W, R, D, S, A, R, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1906</th>\n",
       "      <td>Tatlockia_micdadei|SAMN04274784|GCA_001467875....</td>\n",
       "      <td>IcmT</td>\n",
       "      <td>Leg_IcmT</td>\n",
       "      <td>1.100000e-52</td>\n",
       "      <td>GCA_001467875.1</td>\n",
       "      <td>Tatlockia micdadei</td>\n",
       "      <td>(M, A, G, F, S, E, T, S, H, W, R, D, S, A, R, ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>287 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             targetName queryName  \\\n",
       "2829  Coxiella_burnetii_RSA_493|SAMN02603970|GCA_000...      DotA   \n",
       "4     Fluoribacter_dumoffii_NY_23|SAMN04274771|GCA_0...      DotA   \n",
       "10    Legionella_anisa|SAMN04274764|GCA_001467525.1|...      DotA   \n",
       "8     Legionella_erythra|SAMN04274772|GCA_001467615....      DotA   \n",
       "37    Legionella_geestiana|SAMEA103899863|GCA_900452...      DotA   \n",
       "...                                                 ...       ...   \n",
       "4634  Pseudomonas_monteilii|SAMN11579699|GCA_0098313...      IcmT   \n",
       "4632  Pseudomonas_parafulva_NBRC_16636___DSM_17004|S...      IcmT   \n",
       "4635  Pseudomonas_putida_B001|SAMEA2272788|GCA_00028...      IcmT   \n",
       "4589  Rickettsiella_grylli|SAMN02435840|GCA_00016829...      IcmT   \n",
       "1906  Tatlockia_micdadei|SAMN04274784|GCA_001467875....      IcmT   \n",
       "\n",
       "     queryAccession        E-value              GCA  \\\n",
       "2829       Cox_DotA  1.100000e-280  GCA_000007765.2   \n",
       "4          Leg_DotA   0.000000e+00  GCA_001467605.1   \n",
       "10         Leg_DotA   0.000000e+00  GCA_001467525.1   \n",
       "8          Leg_DotA   0.000000e+00  GCA_001467615.1   \n",
       "37         Leg_DotA   0.000000e+00  GCA_900452375.1   \n",
       "...             ...            ...              ...   \n",
       "4634       Cox_IcmT   6.500000e-06  GCA_009831395.1   \n",
       "4632       Cox_IcmT   1.100000e-06  GCA_000730645.1   \n",
       "4635       Cox_IcmT   6.500000e-06  GCA_000285395.1   \n",
       "4589       Cox_IcmT   1.100000e-43  GCA_000168295.1   \n",
       "1906       Leg_IcmT   1.100000e-52  GCA_001467875.1   \n",
       "\n",
       "                                               tax  \\\n",
       "2829                     Coxiella burnetii RSA 493   \n",
       "4                      Fluoribacter dumoffii NY 23   \n",
       "10                                Legionella anisa   \n",
       "8                               Legionella erythra   \n",
       "37                            Legionella geestiana   \n",
       "...                                            ...   \n",
       "4634                         Pseudomonas monteilii   \n",
       "4632  Pseudomonas parafulva NBRC 16636 = DSM 17004   \n",
       "4635                       Pseudomonas putida B001   \n",
       "4589                          Rickettsiella grylli   \n",
       "1906                            Tatlockia micdadei   \n",
       "\n",
       "                                                    seq  \n",
       "2829  (M, K, K, L, V, S, S, L, L, A, S, I, S, L, F, ...  \n",
       "4     (M, K, R, L, L, G, T, L, L, L, L, L, F, P, G, ...  \n",
       "10    (M, N, K, L, L, V, T, L, L, L, L, V, C, P, G, ...  \n",
       "8     (M, N, K, I, V, S, T, L, L, M, T, L, F, P, A, ...  \n",
       "37    (M, K, K, Y, I, L, G, L, L, S, A, L, M, P, V, ...  \n",
       "...                                                 ...  \n",
       "4634  (M, A, T, G, R, Y, H, W, S, R, C, A, P, P, A, ...  \n",
       "4632  (M, A, T, G, R, Y, H, W, S, R, S, A, P, P, A, ...  \n",
       "4635  (M, A, T, G, R, Y, H, W, S, R, C, A, P, P, A, ...  \n",
       "4589  (M, S, R, V, P, T, A, A, H, W, R, D, S, A, R, ...  \n",
       "1906  (M, A, G, F, S, E, T, S, H, W, R, D, S, A, R, ...  \n",
       "\n",
       "[287 rows x 7 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "proteins_for_tree = set.intersection(*[set(prots) for prots in hmmsrch_df.groupby('tax').queryName.unique()])\n",
    "proteins_to_exclude = {'IcmH_DotU'}\n",
    "proteins_for_tree = proteins_for_tree.union(proteins_to_exclude)\n",
    "hmmsrch_df = hmmsrch_df[hmmsrch_df.queryName.isin(proteins_for_tree)]\n",
    "hmmsrch_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'DotA': set(),\n",
       " 'DotB': set(),\n",
       " 'DotC': set(),\n",
       " 'DotD': set(),\n",
       " 'DotK': set(),\n",
       " 'IcmB_DotO': set(),\n",
       " 'IcmC': set(),\n",
       " 'IcmE_DotG': set(),\n",
       " 'IcmG_DotF': set(),\n",
       " 'IcmH_DotU': {'Rickettsiella grylli'},\n",
       " 'IcmJ_DotN': set(),\n",
       " 'IcmK_DotH': set(),\n",
       " 'IcmL_DotI': set(),\n",
       " 'IcmO_DotL': set(),\n",
       " 'IcmP_DotM': set(),\n",
       " 'IcmT': set()}"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "species = set(hmmsrch_df.tax.unique())\n",
    "with_geest = {k:set(v).symmetric_difference(species) for k, v in dict(hmmsrch_df.groupby('queryName').tax.unique()).items()}\n",
    "with_geest"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Creating the alignment of DotU in ordet to find the root."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DotU_fasta = hmmsrch_df[hmmsrch_df.queryName == 'IcmH_DotU'].seq.to_list()\n",
    "a_tumafaciens_DotU = SeqIO.read(r'/davidb/yatirsolan/thesis_work/figures/novel_insights/T4SSB_figure/full_genomes/Agrobacterium_tumefaciens/Agrobacterium_tumefaciens_DotU.faa', 'fasta')\n",
    "a_tumafaciens_DotU.name = a_tumafaciens_DotU.name.split('|')[0]\n",
    "a_tumafaciens_DotU.id = a_tumafaciens_DotU.name\n",
    "a_tumafaciens_DotU.description = str()\n",
    "DotU_fasta.append(a_tumafaciens_DotU)\n",
    "fasta_file = os.path.join(working_directory, 'IcmH_DotU.faa')\n",
    "SeqIO.write(DotU_fasta, fasta_file, 'fasta')\n",
    "mafft_cline = MafftCommandline(cmd=computational_tools.mafft_paths(algorithm='linsi'), input=fasta_file)\n",
    "stdout, stderr = mafft_cline()\n",
    "alignment = AlignIO.read(io.StringIO(stdout), 'fasta')\n",
    "AlignIO.write(alignment, os.path.join(working_directory, 'IcmH_DotU.aln'), 'fasta')\n",
    "os.system(f\"{computational_tools.tree_path(algorithm='fast_tree')} {os.path.join(working_directory, 'IcmH_DotU.aln')} > {os.path.join(working_directory, 'IcmH_DotU.nw')}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "hmmsrch_df = hmmsrch_df[~(hmmsrch_df.queryName.isin(proteins_to_exclude))]\n",
    "MSA_list = list()\n",
    "for protein, sub_df in hmmsrch_df.groupby('queryName'):\n",
    "    sub_df = sub_df.sort_values('tax')\n",
    "    fasta_file = os.path.join(working_directory, f'{protein}.faa')\n",
    "    SeqIO.write(sub_df.seq, fasta_file, 'fasta')\n",
    "    mafft_cline = MafftCommandline(cmd=computational_tools.mafft_paths(algorithm='linsi'), input=fasta_file)\n",
    "    stdout, stderr = mafft_cline()\n",
    "    alignment = AlignIO.read(io.StringIO(stdout), 'fasta')\n",
    "    AlignIO.write(alignment, fasta_file.replace('.faa', '.aln'), 'fasta')\n",
    "    MSA_list.append(alignment)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for MSA in MSA_list:\n",
    "    try:\n",
    "        cncatntd_MSA += MSA\n",
    "    except NameError:\n",
    "        cncatntd_MSA = MSA\n",
    "        \n",
    "for MSA in cncatntd_MSA:\n",
    "    MSA.description = str()\n",
    "    \n",
    "concatenated_file = os.path.join(working_directory, 'concatenated.aln')\n",
    "AlignIO.write(cncatntd_MSA, concatenated_file, 'fasta')\n",
    "os.system(f\"{computational_tools.tree_path(algorithm='fast_tree')} {concatenated_file} > {os.path.join(working_directory, 'concatenated.nw')}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def determine_color(label):\n",
    "    if 'Legionella' in label or 'Fluoribacter' in label or 'Tatlockia' in label :\n",
    "        return '#2166ac'\n",
    "    elif 'Coxiella' in label or 'Aquicella' in label or 'Rickettsiella' in label:\n",
    "        return '#1b7837'\n",
    "    elif 'Pseu' in label:\n",
    "        return '#d6557c'\n",
    "    elif 'Piscirickettsia' in label:\n",
    "        return '#9B14F0'\n",
    "    return 'black'\n",
    "\n",
    "def f(label):\n",
    "   start = \"\\x1B[3m\"\n",
    "   end = \"\\x1B[0m\"   \n",
    "   return start + label + end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "tree = toytree.tree(os.path.join(working_directory, 'IcmH_DotU.nw'))\n",
    "tree = toytree.tree(os.path.join(working_directory, 'IcmH_DotU.raxml.nw'))\n",
    "tree = tree.root(names=['Agrobacterium_tumefaciens'])\n",
    "tip_labels = [' '.join(label.split('_')[:2:]).replace('Tatlockia', 'Legionella').replace('Fluoribacter', 'Legionella') for label in tree.get_tip_labels()]\n",
    "tip_labels = [\"<i>{}</i>\".format(label) for label in tip_labels]\n",
    "canvas, axes, mark = \\\n",
    "tree.draw(layout='r',\n",
    "           height=400, \n",
    "           width=500,\n",
    "           tip_labels_align=True,\n",
    "           tip_labels_colors = list(map(determine_color, tree.get_tip_labels())),\n",
    "           tip_labels=tip_labels,\n",
    "           tip_labels_style={'font-size':'11px', '-toyplot-anchor-shift':'10px'},\n",
    "           edge_align_style={'stroke':'grey', 'stroke-width':.5, 'stroke-dasharray':'3,1'}) \n",
    "\n",
    "toyplot.html.render(canvas, os.path.join(working_directory, 'IcmH_DotU_raxml_toytree_italic.html'))\n",
    "# canvas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div class=\"toyplot\" id=\"t3677614ab7d54a78b1737a647f9793d5\" style=\"text-align:center\"><svg class=\"toyplot-canvas-Canvas\" height=\"400.0px\" id=\"t842f0b4e3d0f49a68731559fc4830bc3\" preserveAspectRatio=\"xMidYMid meet\" style=\"background-color:transparent;border-color:#292724;border-style:none;border-width:1.0;fill:rgb(16.1%,15.3%,14.1%);fill-opacity:1.0;font-family:Helvetica;font-size:12px;opacity:1.0;stroke:rgb(16.1%,15.3%,14.1%);stroke-opacity:1.0;stroke-width:1.0\" viewBox=\"0 0 600.0 400.0\" width=\"600.0px\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:toyplot=\"http://www.sandia.gov/toyplot\" xmlns:xlink=\"http://www.w3.org/1999/xlink\"><g class=\"toyplot-coordinates-Cartesian\" id=\"tb2d69a9c632e4b319ce35250490a7e92\"><clipPath id=\"tab40bb7da33f475792e47ea30b3d2d10\"><rect height=\"340.0\" width=\"540.0\" x=\"30.0\" y=\"30.0\"></rect></clipPath><g clip-path=\"url(#tab40bb7da33f475792e47ea30b3d2d10)\"><g class=\"toytree-mark-Toytree\" id=\"t15af5f4445e94718aa477ea64bb41f43\"><g class=\"toytree-Edges\" style=\"fill:none;stroke:rgb(14.9%,14.9%,14.9%);stroke-linecap:round;stroke-opacity:1;stroke-width:2\"><path d=\"M 50.0 107.0 L 50.0 69.8 L 114.7 69.8\" id=\"34,33\"></path><path d=\"M 50.0 107.0 L 50.0 144.3 L 114.7 144.3\" id=\"34,32\"></path><path d=\"M 114.7 144.3 L 114.7 108.4 L 370.5 108.4\" id=\"32,31\"></path><path d=\"M 114.7 144.3 L 114.7 180.2 L 143.6 180.2\" id=\"32,30\"></path><path d=\"M 370.5 108.4 L 370.5 122.7 L 384.1 122.7\" id=\"31,29\"></path><path d=\"M 143.6 180.2 L 143.6 201.0 L 154.4 201.0\" id=\"30,28\"></path><path d=\"M 384.1 122.7 L 384.1 134.9 L 389.9 134.9\" id=\"29,27\"></path><path d=\"M 154.4 201.0 L 154.4 226.5 L 194.5 226.5\" id=\"28,26\"></path><path d=\"M 194.5 226.5 L 194.5 261.0 L 198.0 261.0\" id=\"26,25\"></path><path d=\"M 198.0 261.0 L 198.0 222.4 L 200.5 222.4\" id=\"25,24\"></path><path d=\"M 198.0 261.0 L 198.0 299.7 L 205.5 299.7\" id=\"25,23\"></path><path d=\"M 200.5 222.4 L 200.5 236.6 L 204.4 236.6\" id=\"24,22\"></path><path d=\"M 205.5 299.7 L 205.5 281.4 L 209.3 281.4\" id=\"23,21\"></path><path d=\"M 205.5 299.7 L 205.5 318.0 L 216.0 318.0\" id=\"23,20\"></path><path d=\"M 204.4 236.6 L 204.4 248.8 L 208.9 248.8\" id=\"22,19\"></path><path d=\"M 216.0 318.0 L 216.0 330.2 L 219.8 330.2\" id=\"20,18\"></path><path d=\"M 114.7 69.8 L 114.7 61.6 L 134.8 61.6\" id=\"33,17\"></path><path d=\"M 114.7 69.8 L 114.7 77.9 L 131.2 77.9\" id=\"33,16\"></path><path d=\"M 370.5 108.4 L 370.5 94.2 L 389.0 94.2\" id=\"31,15\"></path><path d=\"M 384.1 122.7 L 384.1 110.5 L 394.1 110.5\" id=\"29,14\"></path><path d=\"M 389.9 134.9 L 389.9 126.7 L 392.9 126.7\" id=\"27,13\"></path><path d=\"M 389.9 134.9 L 389.9 143.0 L 390.4 143.0\" id=\"27,12\"></path><path d=\"M 143.6 180.2 L 143.6 159.3 L 220.5 159.3\" id=\"30,11\"></path><path d=\"M 154.4 201.0 L 154.4 175.6 L 226.6 175.6\" id=\"28,10\"></path><path d=\"M 194.5 226.5 L 194.5 191.9 L 213.2 191.9\" id=\"26,9\"></path><path d=\"M 200.5 222.4 L 200.5 208.1 L 231.2 208.1\" id=\"24,8\"></path><path d=\"M 204.4 236.6 L 204.4 224.4 L 219.2 224.4\" id=\"22,7\"></path><path d=\"M 208.9 248.8 L 208.9 240.7 L 221.5 240.7\" id=\"19,6\"></path><path d=\"M 208.9 248.8 L 208.9 257.0 L 222.2 257.0\" id=\"19,5\"></path><path d=\"M 209.3 281.4 L 209.3 273.3 L 220.2 273.3\" id=\"21,4\"></path><path d=\"M 209.3 281.4 L 209.3 289.5 L 219.4 289.5\" id=\"21,3\"></path><path d=\"M 216.0 318.0 L 216.0 305.8 L 226.9 305.8\" id=\"20,2\"></path><path d=\"M 219.8 330.2 L 219.8 322.1 L 225.8 322.1\" id=\"18,1\"></path><path d=\"M 219.8 330.2 L 219.8 338.4 L 224.2 338.4\" id=\"18,0\"></path></g><g class=\"toytree-AlignEdges\" style=\"stroke:rgb(50.2%,50.2%,50.2%);stroke-dasharray:3,1;stroke-linecap:round;stroke-opacity:1.0;stroke-width:0.5\"><path d=\"M 394.1 338.4 L 224.2 338.4\"></path><path d=\"M 394.1 322.1 L 225.8 322.1\"></path><path d=\"M 394.1 305.8 L 226.9 305.8\"></path><path d=\"M 394.1 289.5 L 219.4 289.5\"></path><path d=\"M 394.1 273.3 L 220.2 273.3\"></path><path d=\"M 394.1 257.0 L 222.2 257.0\"></path><path d=\"M 394.1 240.7 L 221.5 240.7\"></path><path d=\"M 394.1 224.4 L 219.2 224.4\"></path><path d=\"M 394.1 208.1 L 231.2 208.1\"></path><path d=\"M 394.1 191.9 L 213.2 191.9\"></path><path d=\"M 394.1 175.6 L 226.6 175.6\"></path><path d=\"M 394.1 159.3 L 220.5 159.3\"></path><path d=\"M 394.1 143.0 L 390.4 143.0\"></path><path d=\"M 394.1 126.7 L 392.9 126.7\"></path><path d=\"M 394.1 110.5 L 394.1 110.5\"></path><path d=\"M 394.1 94.2 L 389.0 94.2\"></path><path d=\"M 394.1 77.9 L 131.2 77.9\"></path><path d=\"M 394.1 61.6 L 134.8 61.6\"></path></g><g class=\"toytree-Nodes\" style=\"fill:rgb(83.5%,24.3%,31%);fill-opacity:1.0;stroke:None;stroke-width:1\"><g id=\"node-0\" transform=\"translate(224.225,338.376)\"><circle r=\"0.0\"></circle></g><g id=\"node-1\" transform=\"translate(225.826,322.097)\"><circle r=\"0.0\"></circle></g><g id=\"node-2\" transform=\"translate(226.918,305.817)\"><circle r=\"0.0\"></circle></g><g id=\"node-3\" transform=\"translate(219.423,289.538)\"><circle r=\"0.0\"></circle></g><g id=\"node-4\" transform=\"translate(220.229,273.258)\"><circle r=\"0.0\"></circle></g><g id=\"node-5\" transform=\"translate(222.190,256.979)\"><circle r=\"0.0\"></circle></g><g id=\"node-6\" transform=\"translate(221.502,240.699)\"><circle r=\"0.0\"></circle></g><g id=\"node-7\" transform=\"translate(219.248,224.419)\"><circle r=\"0.0\"></circle></g><g id=\"node-8\" transform=\"translate(231.240,208.140)\"><circle r=\"0.0\"></circle></g><g id=\"node-9\" transform=\"translate(213.233,191.860)\"><circle r=\"0.0\"></circle></g><g id=\"node-10\" transform=\"translate(226.638,175.581)\"><circle r=\"0.0\"></circle></g><g id=\"node-11\" transform=\"translate(220.465,159.301)\"><circle r=\"0.0\"></circle></g><g id=\"node-12\" transform=\"translate(390.367,143.021)\"><circle r=\"0.0\"></circle></g><g id=\"node-13\" transform=\"translate(392.874,126.742)\"><circle r=\"0.0\"></circle></g><g id=\"node-14\" transform=\"translate(394.143,110.462)\"><circle r=\"4.0\"></circle></g><g id=\"node-15\" transform=\"translate(388.967,94.183)\"><circle r=\"4.0\"></circle></g><g id=\"node-16\" transform=\"translate(131.226,77.903)\"><circle r=\"0.0\"></circle></g><g id=\"node-17\" transform=\"translate(134.756,61.624)\"><circle r=\"0.0\"></circle></g><g id=\"node-18\" transform=\"translate(219.833,330.237)\"><circle r=\"0.0\"></circle></g><g id=\"node-19\" transform=\"translate(208.914,248.839)\"><circle r=\"0.0\"></circle></g><g id=\"node-20\" transform=\"translate(216.041,318.027)\"><circle r=\"0.0\"></circle></g><g id=\"node-21\" transform=\"translate(209.274,281.398)\"><circle r=\"0.0\"></circle></g><g id=\"node-22\" transform=\"translate(204.424,236.629)\"><circle r=\"0.0\"></circle></g><g id=\"node-23\" transform=\"translate(205.532,299.712)\"><circle r=\"0.0\"></circle></g><g id=\"node-24\" transform=\"translate(200.520,222.384)\"><circle r=\"0.0\"></circle></g><g id=\"node-25\" transform=\"translate(197.965,261.048)\"><circle r=\"0.0\"></circle></g><g id=\"node-26\" transform=\"translate(194.490,226.454)\"><circle r=\"0.0\"></circle></g><g id=\"node-27\" transform=\"translate(389.925,134.882)\"><circle r=\"0.0\"></circle></g><g id=\"node-28\" transform=\"translate(154.422,201.017)\"><circle r=\"0.0\"></circle></g><g id=\"node-29\" transform=\"translate(384.056,122.672)\"><circle r=\"0.0\"></circle></g><g id=\"node-30\" transform=\"translate(143.617,180.159)\"><circle r=\"0.0\"></circle></g><g id=\"node-31\" transform=\"translate(370.487,108.427)\"><circle r=\"0.0\"></circle></g><g id=\"node-32\" transform=\"translate(114.748,144.293)\"><circle r=\"0.0\"></circle></g><g id=\"node-33\" transform=\"translate(114.748,69.763)\"><circle r=\"0.0\"></circle></g><g id=\"node-34\" transform=\"translate(50.000,107.028)\"><circle r=\"0.0\"></circle></g></g><g class=\"toytree-TipLabels\" style=\"fill:rgb(14.9%,14.9%,14.9%);fill-opacity:1.0;font-family:helvetica;font-size:11px;font-weight:normal;stroke:none;white-space:pre\"><g transform=\"translate(394.14,338.38)rotate(0)\"><text style=\"fill:rgb(12.9%,40%,67.5%);fill-opacity:1.0\" x=\"10.00\" y=\"2.81\">Legionella dumoffii</text></g><g transform=\"translate(394.14,322.10)rotate(0)\"><text style=\"fill:rgb(12.9%,40%,67.5%);fill-opacity:1.0\" x=\"10.00\" y=\"2.81\">Legionella anisa</text></g><g transform=\"translate(394.14,305.82)rotate(0)\"><text style=\"fill:rgb(12.9%,40%,67.5%);fill-opacity:1.0\" x=\"10.00\" y=\"2.81\">Legionella longbeachae</text></g><g transform=\"translate(394.14,289.54)rotate(0)\"><text style=\"fill:rgb(12.9%,40%,67.5%);fill-opacity:1.0\" x=\"10.00\" y=\"2.81\">Legionella moravica</text></g><g transform=\"translate(394.14,273.26)rotate(0)\"><text style=\"fill:rgb(12.9%,40%,67.5%);fill-opacity:1.0\" x=\"10.00\" y=\"2.81\">Legionella pneumophila</text></g><g transform=\"translate(394.14,256.98)rotate(0)\"><text style=\"fill:rgb(12.9%,40%,67.5%);fill-opacity:1.0\" x=\"10.00\" y=\"2.81\">Legionella micdadei</text></g><g transform=\"translate(394.14,240.70)rotate(0)\"><text style=\"fill:rgb(12.9%,40%,67.5%);fill-opacity:1.0\" x=\"10.00\" y=\"2.81\">Legionella hackeliae</text></g><g transform=\"translate(394.14,224.42)rotate(0)\"><text style=\"fill:rgb(12.9%,40%,67.5%);fill-opacity:1.0\" x=\"10.00\" y=\"2.81\">Legionella erythra</text></g><g transform=\"translate(394.14,208.14)rotate(0)\"><text style=\"fill:rgb(12.9%,40%,67.5%);fill-opacity:1.0\" x=\"10.00\" y=\"2.81\">Legionella geestiana</text></g><g transform=\"translate(394.14,191.86)rotate(0)\"><text style=\"fill:rgb(12.9%,40%,67.5%);fill-opacity:1.0\" x=\"10.00\" y=\"2.81\">Legionella oakridgensis</text></g><g transform=\"translate(394.14,175.58)rotate(0)\"><text style=\"fill:rgb(10.6%,47.1%,21.6%);fill-opacity:1.0\" x=\"10.00\" y=\"2.81\">Rickettsiella grylli</text></g><g transform=\"translate(394.14,159.30)rotate(0)\"><text style=\"fill:rgb(10.6%,47.1%,21.6%);fill-opacity:1.0\" x=\"10.00\" y=\"2.81\">Coxiella burnetii</text></g><g transform=\"translate(394.14,143.02)rotate(0)\"><text style=\"fill:rgb(83.9%,33.3%,48.6%);fill-opacity:1.0\" x=\"10.00\" y=\"2.81\">Pseudomonas monteilii</text></g><g transform=\"translate(394.14,126.74)rotate(0)\"><text style=\"fill:rgb(83.9%,33.3%,48.6%);fill-opacity:1.0\" x=\"10.00\" y=\"2.81\">Pseudomonas putida</text></g><g transform=\"translate(394.14,110.46)rotate(0)\"><text style=\"fill:rgb(83.9%,33.3%,48.6%);fill-opacity:1.0\" x=\"10.00\" y=\"2.81\">Pseudomonas parafulva</text></g><g transform=\"translate(394.14,94.18)rotate(0)\"><text style=\"fill:rgb(83.9%,33.3%,48.6%);fill-opacity:1.0\" x=\"10.00\" y=\"2.81\">Pseudomonas amygdali</text></g><g transform=\"translate(394.14,77.90)rotate(0)\"><text style=\"fill:rgb(60.8%,7.84%,94.1%);fill-opacity:1.0\" x=\"10.00\" y=\"2.81\">Piscirickettsia litoralis</text></g><g transform=\"translate(394.14,61.62)rotate(0)\"><text style=\"fill:rgb(60.8%,7.84%,94.1%);fill-opacity:1.0\" x=\"10.00\" y=\"2.81\">Piscirickettsia salmonis</text></g></g></g></g></g></svg><div class=\"toyplot-behavior\"><script>(function()\n",
       "{\n",
       "var modules={};\n",
       "})();</script></div></div>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "tree = toytree.tree(os.path.join(working_directory, 'concatenated.nw'))\n",
    "tree = toytree.tree(os.path.join(working_directory, 'concatenated.raxml.nw'))\n",
    "tree = tree.root(names=list(filter(lambda tip:'Piscirickettsia' in tip, tree.get_tip_labels())))\n",
    "novel_species = dict(filter(lambda a:a[1] in ['Pseudomonas_amygdali', 'Pseudomonas_parafulva'], tree.get_node_dict().items()))\n",
    "canvas, axes, mark = \\\n",
    "tree.draw(layout='r',\n",
    "           height=400, \n",
    "           width=600,\n",
    "           node_colors='#d53e4f',\n",
    "           tip_labels_align=True,\n",
    "           tip_labels_colors = list(map(determine_color, tree.get_tip_labels())),\n",
    "           tip_labels=[' '.join(label.split('_')[:2:]).replace('Tatlockia', 'Legionella').replace('Fluoribacter', 'Legionella') for label in tree.get_tip_labels()],\n",
    "           node_sizes=[8 if i in novel_species.keys() else 0 for i in tree.get_node_values('idx', True, True)], # size of dash, spacing of dashes\n",
    "           tip_labels_style={'font-size': '11px', '-toyplot-anchor-shift': '10px'},\n",
    "           edge_align_style={'stroke': 'grey', 'stroke-width': .5, 'stroke-dasharray': '3,1'}) \n",
    "\n",
    "toyplot.svg.render(canvas, os.path.join(working_directory, 'figure5_a.svg'))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ddocent_env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
